### PR TITLE
Use contingency context in operator strategies

### PR DESCRIPTION
**Pull request by geofjamg**
Signed-off-by: Geoffroy Jamgotchian <geoffroy.jamgotchian@gmail.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
We can only associate an operator strategy to one contingency.

**What is the new behavior (if this is a feature change)?**
We can thanks to contingency context associate an operator strategy to not only one contingency but also to non or all contingencies.

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: security-analysis/security-analysis-api/src/main/java/com/powsybl/security/json/OperatorStrategyDeserializer.java [hunk 1]
```diff
@@ -73,7 +79,9 @@ public OperatorStrategy deserialize(JsonParser parser, DeserializationContext de
                     return false;
             }
         });
-        OperatorStrategy operatorStrategy = new OperatorStrategy(context.id, context.contingencyId, context.condition, context.actionIds);
+        ContingencyContext contingencyContext = new ContingencyContext(context.contingencyId,
+                context.contingencyContextType != null ? context.contingencyContextType : ContingencyContextType.SPECIFIC);
```
Comment (jeandemanged) [inline security-analysis/security-analysis-api/src/main/java/com/powsybl/security/json/OperatorStrategyDeserializer.java:83]: maybe just add a note that this is done in this way to support both V1.0 and V1.1 JSON format - could be easily forgotten a couple of months from now ...

File: security-analysis/security-analysis-api/src/main/java/com/powsybl/security/json/OperatorStrategyListDeserializer.java [hunk 2]
```diff
@@ -49,8 +49,8 @@ public OperatorStrategyList deserialize(JsonParser parser, DeserializationContex
                     return false;
             }
         });
-        if (context.version == null || !context.version.equals(OperatorStrategyList.VERSION)) {
-            throw new JsonMappingException(parser, "version is missing or not equal to 1.0");
+        if (context.version == null) {
+            throw new JsonMappingException(parser, "version is missing");
```
Comment (jeandemanged) [inline security-analysis/security-analysis-api/src/main/java/com/powsybl/security/json/OperatorStrategyListDeserializer.java:53]: maybe maintain the list of supported versions in OperatorStrategyList, and here continue checking if the version being read is supported ?

Comment (geofjamg) [reply security-analysis/security-analysis-api/src/main/java/com/powsybl/security/json/OperatorStrategyListDeserializer.java:53]: We have 2 problems to implement it like this (correctly indeed):
 - `OperatorStrategy` is used in 2 contexts: as an input (`OperatorStrategyList`) and as an output (inside `SecurityAnalysisResult`). Each of these have their own version and so how could we check version in the jackson deserializer (`OperatorStrategyDeserializer`)? Using which version?
  - there is a bug in jackson 2 that forbid using serialization/deserialization context to carry the version from root deserializer (`SecurityAnalysisResultDeserializer` or `OperatorStrategyListDeserializer`) to children deserializers (like `OperatorStrategyDeserializer`): https://github.com/FasterXML/jackson-databind/issues/1991

File: security-analysis/security-analysis-api/src/main/java/com/powsybl/security/json/OperatorStrategyDeserializer.java (hunk 1 above)
Comment (geofjamg) [reply security-analysis/security-analysis-api/src/main/java/com/powsybl/security/json/OperatorStrategyDeserializer.java:86]: Done

**Review by jeandemanged — APPROVED** [review]

**Review by flo-dup — APPROVED** [review]

**Review by miovd — APPROVED** [review]

**Pull request merged**